Sources/MapboxMaps/Location/Puck/Puck3D.swift

```diff
@@ -89,15 +95,69 @@ internal final class Puck3D: Puck {
             var modelLayer = ModelLayer(id: Self.layerID)
             modelLayer.source = Self.sourceID
             assert(modelLayer.paint != nil)
             modelLayer.paint?.modelLayerType = .constant(.locationIndicator)
-            modelLayer.paint?.modelScale = configuration.modelScale
             modelLayer.paint?.modelRotation = configuration.modelRotation
+            updateModelScale(layer: &modelLayer)
             try! style.addPersistentLayer(modelLayer, layerPosition: nil)
         }
     }
+
+    private func maybeUpdateMercatorScale(at latitude: Double) {
+        let validLatitudeRange = -85.051128779806604...85.051128779806604
+        // In Mercator projection the scale factor is changed along the meridians as a function of latitude
+        // to keep the scale factor equal in all direction: k=sec(latitude), where sec(α) = 1 / cos(α).
+        // Here we are inverting the logic, as the 3D puck is using real-world size, and we are revising
+        // the appearance to look constant on a mercator projection map.
+        let newMercatorScale = cos(latitude.clamped(to: validLatitudeRange) * .pi / 180.0)
+
+        // Threshold to update the mercator scale factor when the latitude changes,
+        // so that we don't update the scale expression too frequently and cause performance issues.
+        if abs(newMercatorScale - mercatorScale) > 0.01 {
+            mercatorScale = newMercatorScale
+        }
+    }
+
+    private func updateModelScale(layer: inout ModelLayer) {
+        let modelScale: Value<[Double]>? = {
+            switch configuration.modelScale {
+            case .constant(let scales):
+                let maxZoom = 22.0
+                let minZoom = 0.5
+                // To make the 3D puck's size constant across different zoom levels, the 3D puck's size (real world object size)
+                // should be exponential to the zoom level.
+                // The base of the exponential expression is decided by how the tile pyramid works:
+                // at zoom level n, we have 2^(n+1) tiles to cover the earth.
+                let exponentialBase = 0.5
+                return .expression(
+                    Exp(.interpolate) {
+                        Exp(.exponential) { exponentialBase }
+                        Exp(.zoom)
+                        minZoom
+                        Exp(.literal) {
+                            scales.map { scale -> Double in
+                                let modelScale = pow(2.0, maxZoom - minZoom)
+                                return modelScale * scale * mercatorScale
+                            }
+                        }
+                        maxZoom
+                        Exp(.literal) {
+                            scales.map { $0 * mercatorScale }
+                        }
+                    }
+                )
+
+            default: return configuration.modelScale
+            }
+        }()
+
+        layer.paint?.modelScale = modelScale
```

Comment:
this is not applied immediately, right? you need to set the whole layer(including other properties) to the style afterwards?

Reply:
I think this might be the reason that the model is not scaled as rapidly as Android, you can set the model scale property separately from the rest of the model layer properties, using `style.setStyleLayerProperty(layerId, "model-scale", scaleExpression)`. This way, the UI probably should look smoother.

Reply:
> this is not applied immediately, right?

Whether true or not, I think the fact that it's not clear suggests that some refactoring is in order to make the flow easier to follow.

> you can set the model scale property separately from the rest of the model layer properties, using style.setStyleLayerProperty(layerId, "model-scale", scaleExpression)

+1. I'd like to see a "create or update" pattern similar to how the source is managed (see lines 85–90 of original)